Add `round_ties_even` to `f32` and `f64`

Jules-Bertholet opened the pull request:
Tracking issue: #96710

Redux of #82273. See also #55107

Adds a new method, `round_ties_even`, to `f32` and `f64`, that rounds the float to the nearest integer , rounding halfway cases to the number with an even least significant bit. Uses the `roundeven` LLVM intrinsic to do this.

Of the five IEEE 754 rounding modes, this is the only one that doesn't already have a round-to-integer function exposed by Rust (others are `round`, `floor`, `ceil`, and `trunc`).  Ties-to-even is also the rounding mode used for int-to-float and float-to-float `as` casts, as well as float arithmentic operations. So not having an explicit rounding method for it seems like an oversight.

Bikeshed: this PR currently uses `round_ties_even` for the name of the method. But maybe `round_ties_to_even` is better, or `round_even`, or `round_to_even`?

Jules-Bertholet committed:
Add `round_ties_even` to `f32` and `f64`

rust-highfive commented:
Some changes occured to rustc_codegen_gcc

cc @antoyo

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @yaahc (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

bors commented:
:umbrella: The latest upstream changes (presumably #95552) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5795171641?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/3a23dfc90a1bf0af64ede8531902e6fadc1a7b1f/checks/5795171641/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |
32  | impl f32 {
    |      ^^^
    |
    = help: consider moving this inherent impl into `core` if possible
help: alternatively add `#[rustc_allow_incoherent_impl]` to the relevant impl items
    |
    |
119 | /     pub fn round_ties_even(self) -> f32 {
120 | |         unsafe { intrinsics::roundevenf32(self) }
    | |_____^

error[E0390]: cannot define inherent `impl` for primitive types outside of `core`
   --> library/std/src/f64.rs:32:6
   --> library/std/src/f64.rs:32:6
    |
32  | impl f64 {
    |      ^^^
    |
    = help: consider moving this inherent impl into `core` if possible
help: alternatively add `#[rustc_allow_incoherent_impl]` to the relevant impl items
    |
    |
119 | /     pub fn round_ties_even(self) -> f64 {
120 | |         unsafe { intrinsics::roundevenf64(self) }
    | |_____^


error[E0599]: no method named `trunc` found for type `f32` in the current scope
    |
    |
162 |         self - self.trunc()
    |                     ^^^^^ method not found in `f32`

error[E0599]: no method named `copysign` found for type `f32` in the current scope
    |
    |
211 |         if self.is_nan() { Self::NAN } else { 1.0_f32.copysign(self) }
    |                                                       ^^^^^^^^ method not found in `f32`

error[E0599]: no method named `trunc` found for type `f32` in the current scope
    |
    |
291 |         let q = (self / rhs).trunc();
    |                              ^^^^^ method not found in `f32`

error[E0599]: no method named `abs` found for type `f32` in the current scope
    |
    |
327 |         if r < 0.0 { r + rhs.abs() } else { r }
    |                              ^^^ method not found in `f32`

error[E0599]: no method named `ln` found for type `f32` in the current scope
    |
    |
478 |         self.ln() / base.ln()
    |              ^^ method not found in `f32`

error[E0599]: no method named `ln` found for type `f32` in the current scope
    |
    |
478 |         self.ln() / base.ln()
    |                          ^^ method not found in `f32`

error[E0599]: no method named `sin` found for type `f32` in the current scope
    |
    |
775 |         (self.sin(), self.cos())
    |               ^^^ help: there is an associated function with a similar name: `min`

error[E0599]: no method named `cos` found for type `f32` in the current scope
    |
    |
775 |         (self.sin(), self.cos())
    |                           ^^^ method not found in `f32`

error[E0599]: no method named `abs` found for type `f32` in the current scope
    |
    |
908 |         (self.abs() + ((self * self) + 1.0).sqrt()).ln().copysign(self)
    |               ^^^ method not found in `f32`

error[E0599]: no method named `sqrt` found for type `f32` in the current scope
    |
    |
908 |         (self.abs() + ((self * self) + 1.0).sqrt()).ln().copysign(self)
    |                                             ^^^^ method not found in `f32`

error[E0599]: no method named `sqrt` found for type `f32` in the current scope
    |
    |
928 |         if self < 1.0 { Self::NAN } else { (self + ((self * self) - 1.0).sqrt()).ln() }
    |                                                                          ^^^^ method not found in `f32`

error[E0599]: no method named `ln` found for type `f32` in the current scope
    |
    |
928 |         if self < 1.0 { Self::NAN } else { (self + ((self * self) - 1.0).sqrt()).ln() }
    |                                                                                  ^^ method not found in `f32`

error[E0599]: no method named `ln_1p` found for type `f32` in the current scope
    |
    |
948 |         0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
    |                                             ^^^^^ method not found in `f32`

error[E0599]: no method named `trunc` found for type `f64` in the current scope
    |
    |
162 |         self - self.trunc()
    |                     ^^^^^ method not found in `f64`

error[E0599]: no method named `copysign` found for type `f64` in the current scope
    |
    |
211 |         if self.is_nan() { Self::NAN } else { 1.0_f64.copysign(self) }
    |                                                       ^^^^^^^^ method not found in `f64`

error[E0599]: no method named `trunc` found for type `f64` in the current scope
    |
    |
291 |         let q = (self / rhs).trunc();
    |                              ^^^^^ method not found in `f64`

error[E0599]: no method named `abs` found for type `f64` in the current scope
    |
    |
327 |         if r < 0.0 { r + rhs.abs() } else { r }
    |                              ^^^ method not found in `f64`

error[E0599]: no method named `log_wrapper` found for type `f64` in the current scope
    |
    |
454 |         self.log_wrapper(|n| unsafe { intrinsics::logf64(n) })
    |              ^^^^^^^^^^^ method not found in `f64`

error[E0599]: no method named `ln` found for type `f64` in the current scope
    |
    |
478 |         self.ln() / base.ln()
    |              ^^ method not found in `f64`

error[E0599]: no method named `ln` found for type `f64` in the current scope
    |
    |
478 |         self.ln() / base.ln()
    |                          ^^ method not found in `f64`

error[E0599]: no method named `log_wrapper` found for type `f64` in the current scope
    |
    |
498 |         self.log_wrapper(|n| {
    |              ^^^^^^^^^^^ method not found in `f64`

error[E0599]: no method named `log_wrapper` found for type `f64` in the current scope
    |
    |
523 |         self.log_wrapper(|n| unsafe { intrinsics::log10f64(n) })
    |              ^^^^^^^^^^^ method not found in `f64`

error[E0599]: no method named `sin` found for type `f64` in the current scope
    |
    |
777 |         (self.sin(), self.cos())
    |               ^^^ help: there is an associated function with a similar name: `min`

error[E0599]: no method named `cos` found for type `f64` in the current scope
    |
    |
777 |         (self.sin(), self.cos())
    |                           ^^^ method not found in `f64`

error[E0599]: no method named `abs` found for type `f64` in the current scope
    |
    |
910 |         (self.abs() + ((self * self) + 1.0).sqrt()).ln().copysign(self)
    |               ^^^ method not found in `f64`

error[E0599]: no method named `sqrt` found for type `f64` in the current scope
    |
    |
910 |         (self.abs() + ((self * self) + 1.0).sqrt()).ln().copysign(self)
    |                                             ^^^^ method not found in `f64`

error[E0599]: no method named `sqrt` found for type `f64` in the current scope
    |
    |
930 |         if self < 1.0 { Self::NAN } else { (self + ((self * self) - 1.0).sqrt()).ln() }
    |                                                                          ^^^^ method not found in `f64`

error[E0599]: no method named `ln` found for type `f64` in the current scope
    |
    |
930 |         if self < 1.0 { Self::NAN } else { (self + ((self * self) - 1.0).sqrt()).ln() }
    |                                                                                  ^^ method not found in `f64`

error[E0599]: no method named `ln_1p` found for type `f64` in the current scope
    |
    |
950 |         0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
    |                                             ^^^^^ method not found in `f64`
Some errors have detailed explanations: E0390, E0599.
For more information about an error, try `rustc --explain E0390`.
error: could not compile `std` due to 31 previous errors
Build completed unsuccessfully in 0:04:05
```

</details>

pnkfelix commented:
Assuming that the symbol names are correct for the cranelift and gcc codegen backends, the compiler changes here look simple and correct.

so I'd say r=me for the compiler side. I'll let @yaahc speak for the libs side.

clarfonthey commented:
Don't think this should block merging this PR, but just to add my opinion to the mix, I've basically always seen this called `round-even` which would translate to `round_even` as a method name. "round ties to even" or "round ties even" just seems unnecessarily verbose, especially when the documentation should alleviate any confusion on what it does.

yaahc commented:
I'm reassigning this PR because I'm taking a break from the review rotation for a little while. Thank you for your patience.

r? rust-lang/libs-api

scottmcm commented:
Given that we have `div_euclid`, maybe something like `round_banker`?  It's short-ish, and gives good search results.

Other inspiration from wikipedia:
> This variant of the round-to-nearest method is also called convergent rounding, statistician's rounding, Dutch rounding, Gaussian rounding, odd–even rounding,[[6]](https://en.wikipedia.org/wiki/Rounding#cite_note-6) or bankers' rounding.

m-ou-se commented:
> Bikeshed: this PR currently uses`round_ties_even` for the name of the method. But maybe `round_ties_to_even` is better, or `round_even`, or `round_to_even`?

We can leave the bikeshed for later, after adding this as unstable but before stabilization.

Can you [open a tracking issue](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) for this feature? Please include the open question about the name of the function, and the open question about whether this should be the default behaviour of `round()` as was mentioned by @dsprenkels on #82273. Thanks!

Jules-Bertholet commented:
Tracking issue opened.

bors commented:
:umbrella: The latest upstream changes (presumably #96877) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

apiraino commented:
Hello, is the status (waiting for review) of this PR correct? Any pending comments from `T-libs`? Thanks

scottmcm commented:
Looks like this has a r=me for the compiler changes, and a "seems reasonable" from libs-api, so I think this is just waiting on a tracking issue (https://github.com/rust-lang/rust/pull/95317#issuecomment-1117194631) to be mergeable, with bikesheds to happen on that tracking issue.  So I've marked it S-author.

Jules-Bertholet commented:
@scottmcm This has had a tracking issue for months, as stated [here](https://github.com/rust-lang/rust/pull/95317#issuecomment-1117458237), but I forgot to link to it in the PR description. Done now

scottmcm commented on library/std/src/f64.rs:
```diff
@@ -89,6 +93,33 @@ impl f64 {
+    #[cfg(not(bootstrap))]
+    #[cfg_attr(not(bootstrap), rustc_allow_incoherent_impl)]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "round_ties_even", issue = "none")]
```
@Jules-Bertholet The issue number from the tracking number needs to be put in the `issue` property of the attribute here (and the other one), instead of `none`, not just opened.

Once that's done, you can say `@rustbot ready` to update the labels to mark for review.

Jules-Bertholet committed:
Add tracking issue

Jules-Bertholet commented:
@rustbot ready

apiraino commented:
I'll signal that this is blocked on https://github.com/rust-lang/rust/issues/96710

@rustbot label S-blocked

Jules-Bertholet commented:
@apiraino #96710 is the tracking issue for this feature, I don't see in what sense this PR is blocked by it?

apiraino commented:
@Jules-Bertholet apologies for the noise, I must have misunderstood the context of this [comment](https://github.com/rust-lang/rust/pull/95317#issuecomment-1176516410). I was trying to figure out the actual status of this PR (ready for merge/review or not yet) in order to make it progress.

@rustbot label -S-blocked

bors commented:
:umbrella: The latest upstream changes (presumably #96946) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #102306) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

scottmcm commented:
@bors r+ rollup=iffy (affects backends)

bors commented:
:pushpin: Commit fdc95afdadcbb060efc7e3bd4615375e47079475 has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm fdc95afdadcbb060efc7e3bd4615375e47079475 -->
<!-- homu: {"type":"Approved","sha":"fdc95afdadcbb060efc7e3bd4615375e47079475","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fdc95afdadcbb060efc7e3bd4615375e47079475 with merge b942dc07ce415568a27dd577053492ec26e7a831...
<!-- homu: {"type":"BuildStarted","head_sha":"fdc95afdadcbb060efc7e3bd4615375e47079475","merge_sha":"b942dc07ce415568a27dd577053492ec26e7a831"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3145874214/jobs/5114259880)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3145874214/jobs/5114259880","builder_name":"checks-actions"} -->

scottmcm commented:
Looks like a plausibly-legit windows linker error
@bors r-

rust-log-analyzer commented:
The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3145874214/jobs/5113687401) [(plain)](https://github.com/rust-lang-ci/rust/commit/b942dc07ce415568a27dd577053492ec26e7a831/checks/8602398389/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- src\f32.rs - f32::f32::round_ties_even (line 101) stdout ----
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcavxMqC\\symbols.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestQCUCom\\rust_out.rust_out.3add85fa-cgu.0.rcgu.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestQCUCom\\rust_out.5492haoi0u6hxqu4.rcgu.o" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\release\\deps" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\compiler_builtins-51812db6bf4c64b0\\out" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\profiler_builtins-4e185bfec47e94b6\\out" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\libstd-4cde99d3445f1df4.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-297e2491bcfb8fbe.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-c480a0eabf04c8cb.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-39a9f701a2d259e2.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-81f544f9f0e43b0f.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-84a39cc58b9dc8a2.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libminiz_oxide-45faf690652fe18a.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libadler-baacf459997bd84f.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-d2480d64abe2992a.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-e81448a99e06d322.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-c971b63f4d1bb1d7.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-9465ef77957dd0fb.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-db8917cca783901c.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-4f1f79de8cc44aa9.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-c5f000ceda76d9c0.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-37bd5d701b9c5d5c.rlib" "kernel32.lib" "advapi32.lib" "userenv.lib" "kernel32.lib" "ws2_32.lib" "bcrypt.lib" "msvcrt.lib" "legacy_stdio_definitions.lib" "/NXCOMPAT" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestQCUCom\\rust_out" "/OPT:REF,NOICF" "/DEBUG"
  = note: rust_out.rust_out.3add85fa-cgu.0.rcgu.o : error LNK2019: unresolved external symbol roundevenf referenced in function _ZN3std3f3221_$LT$impl$u20$f32$GT$15round_ties_even17hb550b4b60bc947a0E
          C:\Users\RUNNER~1\AppData\Local\Temp\rustdoctestQCUCom\rust_out : fatal error LNK1120: 1 unresolved externals

error: aborting due to previous error

Couldn't compile the test.
Couldn't compile the test.
---- src\f64.rs - f64::f64::round_ties_even (line 101) stdout ----
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustcf2qM2b\\symbols.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestt8E2X8\\rust_out.rust_out.3add85fa-cgu.0.rcgu.o" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestt8E2X8\\rust_out.5492haoi0u6hxqu4.rcgu.o" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\release\\deps" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\compiler_builtins-51812db6bf4c64b0\\out" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\build\\profiler_builtins-4e185bfec47e94b6\\out" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-std\\x86_64-pc-windows-msvc\\release\\deps\\libstd-4cde99d3445f1df4.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-297e2491bcfb8fbe.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-c480a0eabf04c8cb.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd_detect-39a9f701a2d259e2.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-81f544f9f0e43b0f.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-84a39cc58b9dc8a2.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libminiz_oxide-45faf690652fe18a.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libadler-baacf459997bd84f.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-d2480d64abe2992a.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-e81448a99e06d322.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-c971b63f4d1bb1d7.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-9465ef77957dd0fb.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-db8917cca783901c.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-4f1f79de8cc44aa9.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-c5f000ceda76d9c0.rlib" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-37bd5d701b9c5d5c.rlib" "kernel32.lib" "advapi32.lib" "userenv.lib" "kernel32.lib" "ws2_32.lib" "bcrypt.lib" "msvcrt.lib" "legacy_stdio_definitions.lib" "/NXCOMPAT" "/LIBPATH:D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustdoctestt8E2X8\\rust_out" "/OPT:REF,NOICF" "/DEBUG"
  = note: rust_out.rust_out.3add85fa-cgu.0.rcgu.o : error LNK2019: unresolved external symbol roundeven referenced in function _ZN3std3f6421_$LT$impl$u20$f64$GT$15round_ties_even17h9f8574b70f1f0b2fE
          C:\Users\RUNNER~1\AppData\Local\Temp\rustdoctestt8E2X8\rust_out : fatal error LNK1120: 1 unresolved externals

error: aborting due to previous error

Couldn't compile the test.
Couldn't compile the test.

failures:
    src\f32.rs - f32::f32::round_ties_even (line 101)
    src\f64.rs - f64::f64::round_ties_even (line 101)

test result: FAILED. 1080 passed; 2 failed; 20 ignored; 0 measured; 0 filtered out; finished in 38.22s

error: doctest failed, to rerun pass `-p std --doc`
Build completed unsuccessfully in 0:41:05
make: *** [Makefile:73: ci-subset-1] Error 1
```

</details>

Jules-Bertholet commented:
I'm not sure what the failure could be caused by. The doc-tests failed with a link error, but the non-doc tests pass, which mean they presumably managed to find the `roundeven` function and link to it.

JohnCSimon commented:
ping from triage:
@Jules-Bertholet 

I see you have a build error - maybe ask around for addressing it? 

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

bors commented:
:umbrella: The latest upstream changes (presumably #104083) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ChrisDenton commented:
cc @nikic maybe? Linking to a llvm intrinsic is failing on Windows in a doc test.

nikic commented:
> I have no idea what the failure could be caused by. The doc-tests failed with a link error, but the non-doc tests pass, which mean they presumably managed to find the `roundeven` function and link to it.

The non-doc tests probably all got constant folded.

nikic commented:
So probably this would need an implementation in https://github.com/rust-lang/libm, plus compiler-builtins reexports for platforms that don't have it natively.

Jules-Bertholet committed:
Use rint instead of roundeven

Use rint intrinsic instead of roundeven to impement `round_ties_even`. They do the same thing when rounding mode is default, which Rust assumes.
And `rint` has better platform support.

Keeps `roundeven` around in `core::intrinsics`, it's doing no harm there.

Jules-Bertholet commented:
I've changed this PR to use `rint` instead of `roundeven`, because that is more widely supported than `roundeven` by various `libc`s, and gives the same results in the default rounding mode (which Rust and LLVM) assume. I've kept the `roundeven` intrinsics around though.

Jules-Bertholet committed:
Remove some `cfg(not(bootstrap))`

Jules-Bertholet commented:
@rustbot ready

nagisa commented:
`rint` doesn’t actually guarantee the ties-to-even rounding mode, does it? It’s behaviour depends on the rounding mode. And although Rust probably won't be allowing anything other than the default ties-to-even mode only, it seems like one could still plausibly set a rounding mode in C, call into Rust and then get back incorrect results – something the `roundeven` function would avoid.

FWIW roundeven is a c++23 function, so Microsoft will have to add an implementation, whether they want or not, when implementing that standard.

Jules-Bertholet commented:
Not only does Rust assume the default rounding mode, so do the LLVM floating-point intrinsics that back it.

bors commented:
:umbrella: The latest upstream changes (presumably #104428) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #104702) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #105328) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #105543) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

albertlarsan68 commented:
@rustbot label -A-bootstrap

Aaron1011 commented:
What's the status of this PR?

apiraino commented:
As far as I know, [this comment](https://github.com/rust-lang/rust/pull/95317#issuecomment-1176516410) still has the current status (though I might have missed progress happening in other places)

roblabla commented:
> As far as I know, [this comment](https://github.com/rust-lang/rust/pull/95317#issuecomment-1176516410) still has the current status (though I might have missed progress happening in other places)

The PR now has a tracking issue, which was the only thing blocking merge at that point AFAICT.

Jules-Bertholet commented:
This PR is ready to merge, as far as I am aware.

apiraino commented:
> This PR is ready to merge, as far as I am aware.

cc @m-ou-se (I think)

WaffleLapkin commented:
- The original implementation was approved by `T-compiler` in https://github.com/rust-lang/rust/pull/95317#issuecomment-1099251055.
- The lib side of things was approved in https://github.com/rust-lang/rust/pull/95317#issuecomment-1176516410
- There are concerns about the new implementation https://github.com/rust-lang/rust/pull/95317#issuecomment-1312598887
  - I've added them to the tracking issue

Given all of that I'll merge this

@bors r=pnkfelix,m-ou-se,scottmcm

All future discussions and work can happen on the tracking issue and in follow-up PRs.

WaffleLapkin commented:
@bors ping

WaffleLapkin commented:
@bors ping

WaffleLapkin commented:
@bors ping

bors commented:
:sleepy: I'm awake I'm awake

WaffleLapkin commented:
*insert a joke about bors being bors*

@bors r=pnkfelix,m-ou-se,scottmcm

bors commented:
:pushpin: Commit 371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e has been approved by `pnkfelix,m-ou-se,scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pnkfelix,m-ou-se,scottmcm 371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e -->
<!-- homu: {"type":"Approved","sha":"371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e","approver":"pnkfelix,m-ou-se,scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e with merge 3e029b13a198ebc9621d0df0d8ce52e64585dabf...
<!-- homu: {"type":"BuildStarted","head_sha":"371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e","merge_sha":"3e029b13a198ebc9621d0df0d8ce52e64585dabf"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4330074861/jobs/7561131440)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4330074861/jobs/7561131440","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4330074861/jobs/7561058976) [(plain)](https://github.com/rust-lang-ci/rust/commit/3e029b13a198ebc9621d0df0d8ce52e64585dabf/checks/11761031312/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2618/3025] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.obj
[2619/3025] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/Reproducer.cpp.obj
[2620/3025] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil-driver.cpp.obj
[2621/3025] Linking CXX static library lib\libLLVMBPFCodeGen.a
FAILED: lib/libLLVMBPFCodeGen.a 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E rm -f lib\libLLVMBPFCodeGen.a && D:\a\rust\rust\mingw64\bin\ar.exe qc lib\libLLVMBPFCodeGen.a  lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFAbstractMemberAccess.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFAdjustOpt.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFAsmPrinter.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFCheckAndAdjustIR.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFFrameLowering.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFInstrInfo.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFIRPeephole.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFISelDAGToDAG.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFISelLowering.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFMCInstLower.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFPreserveDIType.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFRegisterInfo.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFSelectionDAGInfo.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFSubtarget.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFTargetMachine.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFMIPeephole.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFMIChecking.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BPFMISimplifyPatchable.cpp.obj lib/Target/BPF/CMakeFiles/LLVMBPFCodeGen.dir/BTFDebug.cpp.obj && D:\a\rust\rust\mingw64\bin\ranlib.exe lib\libLLVMBPFCodeGen.a && cd ."
D:\a\rust\rust\mingw64\bin\ranlib.exe: could not create temporary file whilst writing archive: no more archived files
[2622/3025] Linking CXX static library lib\libLLVMGlobalISel.a
[2623/3025] Linking CXX static library lib\libLLVMMSP430CodeGen.a
[2624/3025] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/BinaryHolder.cpp.obj
[2625/3025] Building CXX object tools/lli/ChildTarget/CMakeFiles/lli-child-target.dir/ChildTarget.cpp.obj
[2625/3025] Building CXX object tools/lli/ChildTarget/CMakeFiles/lli-child-target.dir/ChildTarget.cpp.obj
[2626/3025] Linking CXX static library lib\libLLVMHexagonCodeGen.a
[2627/3025] Linking CXX executable bin\llvm-profdata.exe
[2628/3025] Linking CXX static library lib\libLLVMPasses.a
ninja: build stopped: subcommand failed.
command did not execute successfully, got: exit code: 1


build script failed, must exit now', C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cmake-0.1.48\src\lib.rs:975:5
 finished in 250.789 seconds
Build completed unsuccessfully in 0:07:03
```

</details>

WaffleLapkin commented:
> D:\a\rust\rust\mingw64\bin\ranlib.exe: could not create temporary file whilst writing archive: no more archived files

Looks spurious.
@bors retry

bors commented:
:hourglass: Testing commit 371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e with merge 9f10e462f6256a1db9f38ef3bc46f917dd753c95...
<!-- homu: {"type":"BuildStarted","head_sha":"371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e","merge_sha":"9f10e462f6256a1db9f38ef3bc46f917dd753c95"} -->

WaffleLapkin commented:
This failure doesn't look related either
@bors retry

bors commented:
:hourglass: Testing commit 371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e with merge 0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4...
<!-- homu: {"type":"BuildStarted","head_sha":"371d57084d7c2e2dd9b4ac6b5bb9fcc8b0dacd1e","merge_sha":"0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4352740116/jobs/7609338492)
Approved by: pnkfelix,m-ou-se,scottmcm
Pushing 0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix,m-ou-se,scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4352740116/jobs/7609338492"},"merge_sha":"0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4"} -->

rust-timer commented:
Finished benchmarking commit (0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4): [comparison URL](https://perf.rust-lang.org/compare.html?start=8f9e09ac3ef3fa85d23ad6a0c920d49987144b13&end=0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=8f9e09ac3ef3fa85d23ad6a0c920d49987144b13&end=0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8f9e09ac3ef3fa85d23ad6a0c920d49987144b13&end=0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.3%, 1.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.8%, 0.8%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-2.2%, -0.9%] | 5     |
| All ❌✅ (primary)                 | 1.3%  | [1.3%, 1.3%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8f9e09ac3ef3fa85d23ad6a0c920d49987144b13&end=0a3b557d528dd7c8a88ceca6f7dc0699b89a3ef4&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

rust-log-analyzer commented:
The job **`i686-mingw-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4349614624/jobs/7599483944) [(plain)](https://github.com/rust-lang-ci/rust/commit/9f10e462f6256a1db9f38ef3bc46f917dd753c95/checks/11809106157/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: make failed
status: exit code: 2
command: "make"
--- stdout -------------------------------
make[1]: Entering directory '/d/a/rust/rust/tests/run-make-fulldeps/target-specs'
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.11.2/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.2/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.2.2/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.17.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.362-9/x64/bin:/c/Program Files/ImageMagick-7.1.0-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.9.0/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs  foo.rs --target=my-awesome-platform.json --crate-type=lib --emit=asm
"D:\a\rust\rust/src/etc/cat-and-grep.sh" -v morestack < /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs/foo.s
[[[ begin stdout ]]]
 .text
 .file "foo.91a024fd-cgu.0"
 .section ".note.GNU-stack","",@progbits
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=i686-pc-windows-gnu target=i686-pc-windows-gnu

[[[ end stdout ]]]
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.11.2/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.2/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.2.2/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.17.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.362-9/x64/bin:/c/Program Files/ImageMagick-7.1.0-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.9.0/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs  foo.rs --target=my-invalid-platform.json 2>&1 | "D:\a\rust\rust/src/etc/cat-and-grep.sh" "Error loading target specification"
[[[ begin stdout ]]]
error: Error loading target specification: expected value at line 1 column 1. Run `rustc --print target-list` for a list of built-in targets


[[[ end stdout ]]]
PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.11.2/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.2/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.2.2/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.17.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.362-9/x64/bin:/c/Program Files/ImageMagick-7.1.0-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.9.0/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs  foo.rs --target=my-incomplete-platform.json 2>&1 | "D:\a\rust\rust/src/etc/cat-and-grep.sh" 'Field llvm-target'
[[[ begin stdout ]]]
error: Error loading target specification: Field llvm-target in target specification is required. Run `rustc --print target-list` for a list of built-in targets


[[[ end stdout ]]]
RUST_TARGET_PATH=. PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.11.2/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.2/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.2.2/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.17.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.362-9/x64/bin:/c/Program Files/ImageMagick-7.1.0-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.9.0/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs  foo.rs --target=my-awesome-platform --crate-type=lib --emit=asm
RUST_TARGET_PATH=. PATH="/d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs:D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0-bootstrap-tools/i686-pc-windows-gnu/release/deps:/d/a/rust/rust/build/i686-pc-windows-gnu/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/mingw32/bin:/c/hostedtoolcache/windows/Python/3.11.2/x64/Scripts:/c/hostedtoolcache/windows/Python/3.11.2/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Program Files/MongoDB/Server/5.0/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/tools/zstd:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.9.3/x64:/c/cabal/bin:/c/ghcup/bin:/c/Program Files/dotnet:/c/mysql/bin:/c/Program Files/R/R-4.2.2/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/npm/prefix:/c/hostedtoolcache/windows/go/1.17.13/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.9/x64/bin:/c/tools/kotlinc/bin:/c/hostedtoolcache/windows/Java_Temurin-Hotspot_jdk/8.0.362-9/x64/bin:/c/Program Files/ImageMagick-7.1.0-Q16-HDRI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/PowerShell/7:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/160/DTS/Binn:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/Program Files/TortoiseSVN/bin:/c/Program Files/CMake/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.9.0/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/nodejs:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/Program Files/GitHub CLI:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Program Files/LLVM/bin:/c/Users/runneradmin/.dotnet/tools:/c/Users/runneradmin/.cargo/bin:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs -L /d/a/rust/rust/build/i686-pc-windows-gnu/test/run-make-fulldeps/target-specs/target-specs  foo.rs --target=my-x86_64-unknown-linux-gnu-platform --crate-type=lib --emit=asm
make[1]: Leaving directory '/d/a/rust/rust/tests/run-make-fulldeps/target-specs'
--- stderr -------------------------------
--- stderr -------------------------------
warning: target json file contains unused fields: morestack

warning: `extern` block uses type `[u8; 16]`, which is not FFI-safe
   |
   |
19 |     fn _foo() -> [u8; 16];
   |                  ^^^^^^^^ not FFI-safe
   = help: consider passing a pointer to the array
   = help: consider passing a pointer to the array
   = note: passing raw arrays by value is not FFI-safe
   = note: `#[warn(improper_ctypes)]` on by default
warning: 1 warning emitted


warning: target json file contains unused fields: morestack

warning: `extern` block uses type `[u8; 16]`, which is not FFI-safe
   |
   |
19 |     fn _foo() -> [u8; 16];
   |                  ^^^^^^^^ not FFI-safe
   = help: consider passing a pointer to the array
   = help: consider passing a pointer to the array
   = note: passing raw arrays by value is not FFI-safe
   = note: `#[warn(improper_ctypes)]` on by default
warning: 1 warning emitted


warning: target json file contains unused fields: morestack

warning: `extern` block uses type `[u8; 16]`, which is not FFI-safe
   |
   |
19 |     fn _foo() -> [u8; 16];
   |                  ^^^^^^^^ not FFI-safe
   = help: consider passing a pointer to the array
   = help: consider passing a pointer to the array
   = note: passing raw arrays by value is not FFI-safe
   = note: `#[warn(improper_ctypes)]` on by default
warning: 1 warning emitted


make[1]: *** [Makefile:8: all] Segmentation fault



failures:
failures:
    [run-make] tests\run-make-fulldeps\target-specs

test result: FAILED. 188 passed; 1 failed; 41 ignored; 0 measured; 0 filtered out; finished in 74.99s

Build completed unsuccessfully in 1:00:54
make: *** [Makefile:78: ci-mingw-subset-1] Error 1
```

</details>

saethlin commented:
Why do the tests here use `asset_approx_eq`? It looks to me like at least the existing test cases could be using `assert_eq`, and that any imprecision at all would be a bug, even if it is less than 1e-6, no?

Jules-Bertholet commented:
I just cargo-culted the `round` tests, which use `assert_approx_eq`. It seems the choice was originally made in https://github.com/rust-lang/rust/commit/0e2242f6d62d6d49bf8a2d1860a41273c1fdfa0d
